File: src/examples/poa/servant_manager/servant_locator.cc
//
// servant_locator.cc
//
// Usage: servant_locator
//
//        On startup, the object reference is printed to cout as a
//        stringified IOR. This string should be used as the argument to 
//        eg2_clt (from the <omniORB>/src/examples/echo/ directory).
//
//        A message of 'shutdown' causes the server (you guessed it) to
//        shutdown.
//

#include <iostream.h>
#include <echo.hh>


static CORBA::ORB_var orb;

//////////////////////////////////////////////////////////////////////

class Echo_i : public POA_Echo,
	       public PortableServer::RefCountServantBase
{
public:
  virtual ~Echo_i() { cout << "Echo_i::~Echo_i()" << endl; }

  char* echoString(const char* mesg)
  {
    cout << "Echo_i::echoString(\"" << mesg << "\")" << endl;
    if( !strcmp(mesg, "shutdown") )  orb->shutdown(0);
    return CORBA::string_dup(mesg);
  }
};

//////////////////////////////////////////////////////////////////////

class MyLocator_i : public POA_PortableServer::ServantLocator,
		    public PortableServer::RefCountServantBase
{
public:
  virtual ~MyLocator_i()
  { cout << "MyLocator_i::~MyLocator_i()" << endl; }

  PortableServer::Servant
  preinvoke(const PortableServer::ObjectId& oid,
	    PortableServer::POA_ptr adapter, const char* operation,
	    PortableServer::ServantLocator::Cookie& the_cookie)
  {
    cout << "MyLocator_i::preinvoke()" << endl;
    return new Echo_i;
  }

  void
  postinvoke(const PortableServer::ObjectId& oid,
	     PortableServer::POA_ptr adapter, const char* operation,
	     PortableServer::ServantLocator::Cookie the_cookie,
	     PortableServer::Servant the_servant)
  {
    cout << "MyLocator_i::postinvoke()" << endl;
    the_servant->_remove_ref();
  }
};

//////////////////////////////////////////////////////////////////////

int main(int argc, char** argv)
{
  try {
    orb = CORBA::ORB_init(argc, argv, "omniORB3");

    // Get hold of the root poa, and activate it.
    CORBA::Object_var obj = orb->resolve_initial_references("RootPOA");
    PortableServer::POA_var root_poa = PortableServer::POA::_narrow(obj);
    PortableServer::POAManager_var poa_man = root_poa->the_POAManager();
    poa_man->activate();

    // Create a new poa to hold our echo objects.
    CORBA::PolicyList pl;
    pl.length(2);
    pl[0] = root_poa->create_request_processing_policy(
				   PortableServer::USE_SERVANT_MANAGER);
    pl[1] = root_poa->create_servant_retention_policy(
					      PortableServer::NON_RETAIN);

    PortableServer::POA_var poa = root_poa->create_POA("mypoa", poa_man, pl);

    // Activate a servant locator in the root poa.
    MyLocator_i* sl = new MyLocator_i;
    PortableServer::ObjectId_var id = root_poa->activate_object(sl);
    PortableServer::ServantLocator_var slref = sl->_this();
    sl->_remove_ref();

    // Register the servant activator with our new poa.
    poa->set_servant_manager(slref);

    // Print out a reference to an object.
    CORBA::Object_var ref = poa->create_reference("IDL:Echo:1.0");
    CORBA::String_var sior(orb->object_to_string(ref));
    cout << "'" << (char*) sior << "'" << endl;

    orb->run();
    orb->destroy();
  }
  catch(CORBA::SystemException&) {
    cerr << "Caught CORBA::SystemException." << endl;
  }
  catch(CORBA::Exception&) {
    cerr << "Caught CORBA::Exception." << endl;
  }
  catch(omniORB::fatalException& fe) {
    cerr << "Caught omniORB::fatalException:" << endl;
    cerr << "  file: " << fe.file() << endl;
    cerr << "  line: " << fe.line() << endl;
    cerr << "  mesg: " << fe.errmsg() << endl;
  }
  catch(...) {
    cerr << "Caught unknown exception." << endl;
  }

  return 0;
}

File: omniORBpy/examples/poa/servantlocator.py
#!/usr/bin/env python

# Example of using a ServantLocator.
#

# This program behaves just like the echo server, except that the
# servant object is activated and deactivated again on each request to
# the object.
#
# If you run with a -l command line argument, the program will make a
# number of local calls to the object, showing that the activator runs
# in the local, as well as the remote, case.

import sys, time
from omniORB import CORBA, PortableServer

import _GlobalIDL, _GlobalIDL__POA

class Echo_i (_GlobalIDL__POA.Echo):
    def __init__(self):
        print "Echo_i created."

    def __del__(self):
        print "Echo_i deleted."

    def echoString(self, mesg):
        print "echoString() called with message:", mesg, \
              ", ObjectId =", repr(current.get_object_id())
        return mesg

class ServantLocator_i (PortableServer.ServantLocator):
    def preinvoke(self, oid, poa, operation):
        print "preinvoke(): oid:", oid, "poa:", poa._get_the_name()
        if oid == "MyEcho2":
            raise PortableServer.ForwardRequest(rooteo)

        ei = Echo_i()
        return (ei, "Hmm, cookies")

    def postinvoke(self, oid, poa, operation, cookie, serv):
        print "postinvoke(): oid:", oid, "poa:", poa._get_the_name(), \
              "cookie: '" + cookie + "'"

# Initialise the ORB and activate the root POA.
orb = CORBA.ORB_init(sys.argv, CORBA.ORB_ID)
poa = orb.resolve_initial_references("RootPOA")
poaManager = poa._get_the_POAManager()
poaManager.activate()

current = orb.resolve_initial_references("POACurrent")

# Create a child POA with the right policies for a ServantLocator
ps = [poa.create_id_assignment_policy(PortableServer.USER_ID),
      poa.create_servant_retention_policy(PortableServer.NON_RETAIN),
      poa.create_request_processing_policy(PortableServer.USE_SERVANT_MANAGER)]

child = poa.create_POA("MyPOA", poaManager, ps)

# Create the ServantLocator and set it as the child's ServantManager
sli = ServantLocator_i()
child.set_servant_manager(sli)

rooteo = Echo_i()._this()

# Create an object reference with no servant
eo = child.create_reference_with_id("MyEcho", CORBA.id(_GlobalIDL.Echo))
eo2 = child.create_reference_with_id("MyEcho2", CORBA.id(_GlobalIDL.Echo))
print orb.object_to_string(eo)

# Run, or do some local calls...
if not (len(sys.argv) > 1 and sys.argv[1] == "-l"):
    orb.run()

time.sleep(1)

print "Calling..."

# On each of these calls, the servant is created just before the
# invocation, and deleted again just afterwards

print eo.echoString("Hello from same address space")
time.sleep(1)

print eo.echoString("Hello again")
time.sleep(1)

print eo.echoString("Hello again again")
time.sleep(1)

print eo2.echoString("Hello object 2")
time.sleep(1)

print "Destroying child POA..."
child.destroy(1, 1)
print "Child POA Destroyed"

print "Destroying ORB..."
orb.destroy()
print "ORB destroyed."
